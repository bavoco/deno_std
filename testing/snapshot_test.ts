// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
import { stripColor } from "../fmt/colors.ts";
import { ensureDir } from "../fs/mod.ts";
import { dirname, fromFileUrl, join } from "../path/mod.ts";
import { assert, assertInstanceOf, AssertionError, fail } from "./asserts.ts";
import { assertSnapshot } from "./snapshot.ts";

class TestClass {
  a = 1;
  b = 2;
  init() {
    this.b = 3;
  }
  get getA() {
    return this.a;
  }
  func() {}
}

const map = new Map();
map.set("Hello", "World!");
map.set(() => "Hello", "World!");
map.set(1, 2);

Deno.test("Snapshot Test", async (t) => {
  await assertSnapshot(t, { a: 1, b: 2 });
  await assertSnapshot(t, new TestClass());
  await assertSnapshot(t, map);
  await assertSnapshot(t, new Set([1, 2, 3]));
  await assertSnapshot(t, { fn() {} });
  await assertSnapshot(t, function fn() {});
  await assertSnapshot(t, [1, 2, 3]);
  await assertSnapshot(t, "hello world");
});

Deno.test("Snapshot Test - step", async (t) => {
  await assertSnapshot(t, { a: 1, b: 2 });
  await t.step("Nested", async (t) => {
    await assertSnapshot(t, new TestClass());
    await assertSnapshot(t, map);
    await t.step("Nested Nested", async (t) => {
      await assertSnapshot(t, new Set([1, 2, 3]));
      await assertSnapshot(t, { fn() {} });
      await assertSnapshot(t, function fn() {});
    });
    await assertSnapshot(t, [1, 2, 3]);
  });
  await assertSnapshot(t, "hello world");
});

Deno.test("Snapshot Test - Adverse String \\ ` ${}", async (t) => {
  await assertSnapshot(t, "\\ ` ${}");
});

Deno.test("Snapshot Test - Failed Assertion", async (t) => {
  await t.step("Object", async (t) => {
    try {
      await assertSnapshot(t, [1, 2]);
      fail("Expected snapshot not to match");
    } catch (error) {
      assertInstanceOf(error, AssertionError);
      await assertSnapshot(t, stripColor(error.message).split("\n"));
    }
  });
  await t.step("String", async (t) => {
    try {
      await assertSnapshot(t, "Hello!");
      fail("Expected snapshot not to match");
    } catch (error) {
      assertInstanceOf(error, AssertionError);
      await assertSnapshot(t, stripColor(error.message).split("\n"));
    }
  });
});

Deno.test("Snapshot Test - Update", async (t) => {
  const TEMP_DIR = ".tmp";

  async function runTestWithUpdateFlag(test: string, deleteTempDir = true) {
    const testDir = dirname(fromFileUrl(import.meta.url));
    const tempDir = join(testDir, TEMP_DIR);
    const tempTestFileName = "test.ts";
    const tempTestFilePath = join(tempDir, tempTestFileName);
    await ensureDir(tempDir);
    await Deno.writeTextFile(tempTestFilePath, test);

    const process = await Deno.run({
      cmd: ["deno", "test", "--allow-all", tempTestFilePath, "--", "-u"],
      stdout: "piped",
      stderr: "piped",
    });
    const output = await process.output();
    const error = await process.stderrOutput();
    process.close();

    if (deleteTempDir) {
      await Deno.remove(tempDir, { recursive: true });
    }

    return {
      output: new TextDecoder().decode(output),
      error: new TextDecoder().decode(error),
    };
  }

  function formatOutput(string: string) {
    // Strip colors and obfuscate any timings
    return stripColor(string).replace(/([0-9])+m?s/g, "--ms");
  }

  function formatError(string: string) {
    // Strip colors and remove "Check file:///workspaces/deno_std/testing/.tmp/test.ts"
    // as this is always output to stderr
    return stripColor(string).replace(/^Check file:\/\/(.+)\n/g, "");
  }

  /**
   * New snapshot
   */
  const result1 = await runTestWithUpdateFlag(
    `
    import { assertSnapshot } from "../snapshot.ts";

    Deno.test("Snapshot Test - Update", async (t) => {
      await assertSnapshot(t, [
        1,
        2,
      ]);
    });
    `,
    false,
  );

  await assertSnapshot(t, formatOutput(result1.output).split("\n"));
  assert(!formatError(result1.error), "unexpected output to stderr");

  /**
   * Existing snapshot - no changes
   */
  const result2 = await runTestWithUpdateFlag(
    `
    import { assertSnapshot } from "../snapshot.ts";

    Deno.test("Snapshot Test - Update", async (t) => {
      await assertSnapshot(t, [
        1,
        2,
      ]);
    });
    `,
    false,
  );

  await assertSnapshot(t, formatOutput(result2.output).split("\n"));
  assert(!formatError(result2.error), "unexpected output to stderr");

  /**
   * Existing snapshot - updates
   */
  const result3 = await runTestWithUpdateFlag(`
    import { assertSnapshot } from "../snapshot.ts";

    Deno.test("Snapshot Test - Update", async (t) => {
      await assertSnapshot(t, [
        1,
        2,
        3,
        5,
      ]);
    });
  `);

  await assertSnapshot(t, formatOutput(result3.output).split("\n"));
  assert(!formatError(result3.error), "unexpected output to stderr");
});

// Regression test for https://github.com/denoland/deno_std/issues/2140
// Long strings should not be truncated with ellipsis
Deno.test("Snapshot Test - Regression #2140", async (t) => {
  await assertSnapshot(t, {
    title: "Testing a page",
    content: `
      <h1>Testing a page</h1>
      <p>This is a test</p>
      <ul>
        <li>1</li>
        <li>2</li>
        <li>3</li>
        <li>4</li>
      </ul>
      `,
  });
});

// Regression test for https://github.com/denoland/deno_std/issues/2144
// Empty arrays should be compacted
Deno.test("Snapshot Test - Regression #2144", async (t) => {
  const config = {
    fmt: {
      files: {
        exclude: [],
        include: [],
      },
      options: {},
    },
  };
  await assertSnapshot(t, config);
});
