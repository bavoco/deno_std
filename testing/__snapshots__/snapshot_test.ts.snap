export const snapshot = {};

snapshot[`Snapshot Test 1`] = `
{
  a: 1,
  b: 2,
}
`;

snapshot[`Snapshot Test 2`] = `
TestClass {
  a: 1,
  b: 2,
}
`;

snapshot[`Snapshot Test 3`] = `
Map {
  "Hello" => "World!",
  1 => 2,
  [Function] => "World!",
}
`;

snapshot[`Snapshot Test 4`] = `
Set {
  1,
  2,
  3,
}
`;

snapshot[`Snapshot Test 5`] = `
{
  fn: [Function: fn],
}
`;

snapshot[`Snapshot Test 6`] = `[Function: fn]`;

snapshot[`Snapshot Test 7`] = `
[
  1,
  2,
  3,
]
`;

snapshot[`Snapshot Test 8`] = `"hello world"`;

snapshot[`Snapshot Test - step 1`] = `
{
  a: 1,
  b: 2,
}
`;

snapshot[`Snapshot Test - step > Nested 1`] = `
TestClass {
  a: 1,
  b: 2,
}
`;

snapshot[`Snapshot Test - step > Nested 2`] = `
Map {
  "Hello" => "World!",
  1 => 2,
  [Function] => "World!",
}
`;

snapshot[`Snapshot Test - step > Nested > Nested Nested 1`] = `
Set {
  1,
  2,
  3,
}
`;

snapshot[`Snapshot Test - step > Nested > Nested Nested 2`] = `
{
  fn: [Function: fn],
}
`;

snapshot[`Snapshot Test - step > Nested > Nested Nested 3`] = `[Function: fn]`;

snapshot[`Snapshot Test - step > Nested 3`] = `
[
  1,
  2,
  3,
]
`;

snapshot[`Snapshot Test - step 2`] = `"hello world"`;

snapshot[`Snapshot Test - Adverse String \\ \` \${} 1`] = `"\\\\ \` \${}"`;

snapshot[`Snapshot Test - Failed Assertion > Object 1`] = `
[
  1,
  2,
  3,
]
`;

snapshot[`Snapshot Test - Failed Assertion > Object 2`] = `
[
  "Snapshot does not match:",
  "",
  "",
  "    [Diff] Actual / Expected",
  "",
  "",
  "    [",
  "      1,",
  "      2,",
  "+     3,",
  "    ]",
  "",
]
`;

snapshot[`Snapshot Test - Failed Assertion > String 1`] = `"Hello World!"`;

snapshot[`Snapshot Test - Failed Assertion > String 2`] = `
[
  "Snapshot does not match:",
  "",
  "",
  "    [Diff] Actual / Expected",
  "",
  "",
  '-   "Hello!"',
  '+   "Hello World!"',
  "",
  "",
]
`;

snapshot[`Snapshot Test - Update 1`] = `
[
  "running 1 test from ./testing/.tmp/test.ts",
  "Snapshot Test - Update ... ok (--ms)",
  "",
  " > 1 snapshots updated.",
  "",
  "test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (--ms)",
  "",
  "",
]
`;

snapshot[`Snapshot Test - Update 2`] = `
[
  "running 1 test from ./testing/.tmp/test.ts",
  "Snapshot Test - Update ... ok (--ms)",
  "",
  "test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (--ms)",
  "",
  "",
]
`;

snapshot[`Snapshot Test - Update 3`] = `
[
  "running 1 test from ./testing/.tmp/test.ts",
  "Snapshot Test - Update ... ok (--ms)",
  "",
  " > 1 snapshots updated.",
  "",
  "test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (--ms)",
  "",
  "",
]
`;

snapshot[`Snapshot Test - Regression #2140 1`] = `
{
  content: "\\n      <h1>Testing a page</h1>\\n      <p>This is a test</p>\\n      <ul>\\n        <li>1</li>\\n        <li>2</li>\\n        <li>3</li>\\n        <li>4</li>\\n      </ul>\\n      ",
  title: "Testing a page",
}
`;

snapshot[`Snapshot Test - Regression #2144 1`] = `
{
  fmt: {
    files: {
      exclude: [
      ],
      include: [
      ],
    },
    options: {},
  },
}
`;
